---
title: "Lab 3 - Trying out GWAS"
author: "Alencar Xavier, Gota Morota"
date: "October 26, 2018"
widescreen: true
output: ioslides_presentation
---

# Prelude: Data & Structure

```{r,echo=FALSE}
require(NAM,quietly=TRUE)
```

## Getting some data

Example dataset from the *SoyNAM* package. We are querying two of the forty biparental families with a shared parental IA3023, grown in 18 environment.

```{r}
Data = SoyNAM::BLUP(trait = 'yield', family = 2:3)
```

## Genomic relationship matrix

```{r}
y = Data$Phen
M = Data$Gen
#
Z = apply(M,2,function(snp) snp-mean(snp))
ZZ = tcrossprod(Z)
Sum2pq = sum(apply(M,2,function(snp){p=mean(snp)/2; return(2*p*(1-p))}))
G = ZZ/Sum2pq
```

Kernel commonly deployed, referred in VanRaden (2008)

$$G=\frac{(M-P)(M-p)'}{2\sum^{J}_{j=1}{p_j(1-p_j)}}$$

## Genomic relationship matrix

```{r,fig.width=5,fig.height=5,fig.align="center"}
image(G[,241:1], main='GRM heatmap',xaxt='n',yaxt='n')
```


## Structure parameters (1) PCs

```{r,fig.width=6,fig.height=4,fig.align="center"}
Spectral = eigen(G,symmetric = TRUE)
PCs = Spectral$vectors[,1:5]
plot(PCs,xlab='PC 1',ylab='PC 2',main='Population on eigen spaces',col=Data$Fam,pch=20)
```

## Structure parameters (2) Clusters

```{r,fig.width=6,fig.height=4,fig.align="center"}
GeneticDistance = Gdist(M,method=6)
Tree = hclust(GeneticDistance,method = 'ward.D2')
plot(Tree,labels = FALSE)
Clst = factor(cutree(Tree,k=2))
```

# Single marker analysis

## GLM (1) - No structure

```{r}
Marker = M[,117]
#
fit = lm( y ~ Marker )
anova( fit )
-log(anova(fit)$`Pr(>F)`[1],base = 10)
```

## GLM (2) - Principal Components

```{r}
reduced_model = lm( y ~ PCs )
full_model = lm( y ~ PCs + Marker )
anova( reduced_model, full_model )
-log((anova( reduced_model, full_model ))$`Pr(>F)`[2],base = 10)
```

## GLM (3) - Population Clusters

```{r}
reduced_model = lm( y ~ Clst )
full_model = lm( y ~ Clst + Marker )
anova( reduced_model, full_model )
-log( anova(reduced_model,full_model)$`Pr(>F)`[2],base = 10)
```

## MLM - K+Q model

```{r}
Q = model.matrix(~Clst)
reduced_model = reml( y=y, X=Q, K=G)
full_model = reml( y=y, X=cbind(Q, Marker), K=G)
LRT = -2*(full_model$loglik - reduced_model$loglik)
-log(pchisq(LRT,1,lower.tail=FALSE),base=10)
```

# Whole genome screening

## DYI (example with GLM)

```{r,fig.width=8,fig.height=3.5,fig.align="center"}
reduced_model = lm( y ~ Clst )
glm_pval = apply(M,2,function(Marker){
  pval = anova(reduced_model, lm(y~Clst+Marker) )$`Pr(>F)`[2]
  return(-log(pval,base = 10))})
plot(glm_pval,pch=20,xlab='SNP',main='My first GLM GWAS')
```

## Using CRAN implementations

NAM random model: $y=\mu+Marker \times Pop+Zu+e$ 

```{r}
fit_gwa = gwas3(y=y, gen=M, fam=c(Clst), chr=Data$Chrom)
```

## Manhattan plot

```{r,fig.width=8,fig.height=4,fig.align="center"}
plot(fit_gwa, pch=20, main = "My first MLM GWAS")
```

## Yet another R implementations

```{r}
require(rrBLUP,quietly = TRUE); COL = fit_gwa$MAP[,1]%%2+1 # Color chromosomes
geno=data.frame(colnames(M),fit_gwa$MAP[,1:2],t(M-1),row.names=NULL)
pheno=data.frame(line=colnames(geno)[-c(1:3)],Pheno=y,Clst,row.names=NULL)
fit_another_gwa=GWAS(pheno,geno,fixed='Clst',plot=FALSE)
```

```{r,echo=FALSE,fig.width=6,fig.height=2.5,fig.align="center"}
mlm_pval=fit_another_gwa$Pheno; mlm_pval[mlm_pval==0]=NA
par(mar=c(0,0,0,0))
plot(mlm_pval,col=COL+3,pch=20);
```


## Comparing results

```{r,fig.width=4,fig.height=4,fig.align="center"}
mlm_pval=fit_another_gwa$Pheno; mlm_pval[mlm_pval==0]=NA
plot(glm_pval,mlm_pval,xlab='GLM',ylab='MLM',main='Compare')
```

## Power analysis - QQ plot

```{r,fig.width=9,fig.height=3,fig.align="center"}
nam_pval = fit_gwa$PolyTest$pval
par(mfrow=c(1,3))
qqman::qq(glm_pval,main='GLM')
qqman::qq(mlm_pval,main='MLM')
qqman::qq(nam_pval,main='RLM')
```

# Multiple testing

## Multiple testing

*Wikipedia:* In statistics, the **multiple comparisons**, **multiplicity** or **multiple testing** problem occurs when one considers a set of statistical inferences simultaneously or infers a subset of parameters selected based on the observed values.In certain fields it is known as the look-elsewhere effect: *The more inferences are made, the more likely erroneous inferences are to occur.* Several statistical techniques have been developed to prevent this from happening, allowing significance levels for single and multiple comparisons to be directly compared. These techniques generally require a **stricter significance threshold** for individual comparisons, so as to compensate for the number of inferences being made.

## Baseline - No correction

Base significance threshold: $\alpha=0.05/m$

```{r,fig.width=8,fig.height=4,fig.align="center"}
plot(fit_gwa, alpha=0.05, main = "No correction")
```


## Multiple testing correction

Bonferroni: $\alpha=0.05/m$

```{r,fig.width=8,fig.height=4,fig.align="center"}
plot(fit_gwa, alpha=0.05/ncol(M), main = "Bonferroni correction")
```

## False-Discovery Rate

Benjamini-Hochberg FDR: $\alpha=\frac{0.05}{m\times (1-FDR)}$

```{r,fig.width=8,fig.height=4,fig.align="center"}
plot(fit_gwa, alpha=0.05/(ncol(M)*.75), main = "FDR 25%")
```

## False-Discovery Rate

Unique segments based on Eigenvalues: $m^* = D > 1$

```{r,fig.width=8,fig.height=4,fig.align="center"}
m_star = sum(Spectral$values>1)
plot(fit_gwa, alpha=0.05/m_star, main="Bonferroni on unique segments")
```

# Multi-loci analysis

## Whole genome regression

```{r,fig.width=8,fig.height=4,fig.align="center"}
fit_wgr = bWGR::BayesDpi(y=y,X=M,it=3000); par(mfrow=c(1,2));
plot(fit_wgr$PVAL,col=COL,pch=20,ylab='-log(p)',main='GWA')
plot(fit_wgr$b,col=COL,pch=20,ylab='Marker effect',main='GWP')
plot(fit_wgr$hat,y,pch=20)
```

## WGR - No need for multiple testing

```{r,fig.width=8,fig.height=4,fig.align="center"}
thr_none = -log(pchisq(qchisq(1-0.05/ncol(M),1),1,lower.tail=FALSE),base=10)
thr_bonf = -log(pchisq(qchisq(1-0.05,1),1,lower.tail=FALSE),base=10)
par(mfrow=c(1,2)); plot(fit_gwa,alpha=NULL,main="MLM",pch=20); abline(h=thr_none,col=3)
plot(fit_wgr$PVAL,col=COL,ylab='-log(p)',main="WGR",pch=20); abline(h=thr_bonf,col=3)
```

## Approaches are complementary

```{r,echo=FALSE,fig.width=10,fig.height=5,fig.align="center"}
thr_fdr = -log(pchisq(qchisq(1-0.05/m_star,1),1,lower.tail=FALSE),base=10)
par(mfrow=c(2,2),mar=c(0,1,1.5,1));
plot(glm_pval,main="GLM",col=COL+2,pch=20,,xaxt='n',yaxt='n');
abline(h=thr_none,col=3); abline(h=thr_bonf,lty=2,col=3); abline(h=thr_fdr,lty=3,col=3);
plot(mlm_pval,alpha=NULL,main="MLM",col=COL+1,pch=20,xaxt='n',yaxt='n');
abline(h=thr_none,col=3); abline(h=thr_bonf,lty=2,col=3); abline(h=thr_fdr,lty=3,col=3);
plot(fit_gwa,alpha=NULL,main="RLM",pch=20,xaxt='n',yaxt='n');
abline(h=thr_none,col=3); abline(h=thr_bonf,lty=2,col=3); abline(h=thr_fdr,lty=3,col=3);
plot(fit_wgr$PVAL,col=COL,ylab='-log(p)',main="WGR",pch=20,xaxt='n',yaxt='n');
abline(h=thr_none,col=3); abline(h=thr_bonf,lty=2,col=3); abline(h=thr_fdr,lty=3,col=3);
```

## Random forest

```{r,fig.width=10,fig.height=4,fig.align="center"}
fit_rf = ranger::ranger(y~.,data= data.frame(y=y,M),importance='impurity')
plot(fit_rf$variable.importance,ylab='Importance',main='Random Forest',col=COL+7,pch=20)
```


# Break
