---
title: "Lecture 3 - Basics of GWAS and signal detection"
author: "Alencar Xavier, Gota Morota"
date: "October 26, 2018"
widescreen: true
output: ioslides_presentation
---

<style>
  .col2 {
    columns: 2 200px;
    -webkit-columns: 2 200px;
    -moz-columns: 2 200px;
  }
</style>

## Outline

<div class="col2">
- Test statistics
- Allele coding
- Power & resolution
- Linkage mapping
- LD mapping
- Structure
- Imputation
- GLM
- MLM
- WGR
- Rare-variants
- Validation studies
</div>

```{r,echo=F}
require(NAM,quietly = T)
```


## Test statistics

- Testing associations are as simple as t-test and ANOVA

```{r,echo=F,fig.width=8,fig.height=4,fig.align="center"}
par(mfrow=c(1,2))
set.seed(1)
geno = factor(sort(rep(c(c('AA','Aa','aa')),100)))
pheno = c( rnorm(100,22,4), rnorm(100,24,5), rnorm(100,26,4) )
ppheno = c(pheno[1:200],pheno[101:300])
allele = factor(sort(rep(c(c('A','a')),200)))
boxplot(ppheno~allele,col=rainbow(5)[1:2],ylab='Phenotype',main='Allele frequencies (t-test)')
boxplot(pheno~geno,col=rainbow(5)[3:5],ylab='Phenotype',main='Genotypic frequencies (ANCOVA)')
abline(lm(pheno~as.numeric(geno)),lwd=2,col=2,lty=4)
```

## Test statistics

- A more generalized framework: Likelihood test

$$LRT = L_0 / L_1 = -2(logL_1 - logL_0)$$

For the model:

$$y=Xb+Zu+e\\ y\sim N(Xb,V)$$

REML function is given by:

$$L(\sigma^2_u,\sigma^2_e)=-0.5( ln|V|+ln|X'V^{-1}X|+y'Py)$$

Where $V=ZKZ'\sigma^2_u+I\sigma^2_e$ and $y'Py=y'e$


## Allele coding

Types of allele coding

1. Add. (1 df): {-1,0,1} or {0,1,2} - **Very popular** (Lines, GCA)
2. Dom. (1 df): {0,1,0} - **Popular** (Trees, clonals and Hybrids)
3. Jointly A+D (2 df): Popular on QTL mapping in F2s
4. Complete dominance (1 df): {0,0,1} or {0,1,1} - **Very unusual**
5. Interactions (X df): *Marker x Factor* (epistasis and GxE)

## Power and resolution

**Power**

- Key: Number of individuals & allele frequency
- More DF = lower power
- Multiple testing: Bonferroni and FDR
- Tradeoff: Power vs false positives

**Resolution**

- Genotyping density
- LD blocks
- Recombination

## Power: Variance of X

```{r,echo=F,fig.width=8,fig.height=5,fig.align="center"}
curve(2*x*(1-x),xlab='p',ylab='Var(x)',ylim=c(0,1.1),col=2,lwd=2,main='Marker variance')
curve( 4*x*(1-x), ylim=c(0,.6),col=4,lwd=2, add=T)
#
legend('topright',legend=c('HW (2pq)','Inbred (4pq)'),col=c(2,4),lty=1,lwd=2,bty='n')
#
text(0.5,0.55,'Unselected DH / RIL',col=2)
text(0.5,1.05,'Unselected F2',col=4)
lines(x=c(.5,.5),y=c(.5,1),col=c(2,4),pch=20,lwd=5,type='p')
#
arrows(x0=.5,x1=.5,y0=0,y1=.2,code = 1)
text(0.5,0.25,'p = q',cex=1.5)

```

##  Beavis effect: 1000 is just OK

![](https://raw.githubusercontent.com/alenxav/miscellaneous/master/Beavis_effect.png)

## Multiple testing:

GWAS tests $m$ hypothesis:

- No correction: $\alpha = 0.05/m$
- Bonferroni: $\alpha = 0.05/m$
- FDR (25\%): $\alpha = 0.05/(m\times0.75)$

```{r,echo=FALSE,fig.width=5,fig.height=3,fig.align="center"}
par(mar=c(1,1,1,1))
data(tpod); h=y+emRR(y,gen)$hat; fit = emGWA(h, gen);
CHR = c(); for(i in 1:20) CHR=c(CHR,rep(i,chr[i]))
plot(fit$PVAL*2,pch=16,col=CHR,ylim=c(0,5),xaxt='n',yaxt='n')
abline(h=c(1.2,3.8,3.3),lty=2,col=1:3) 
legend('topleft',c('No correction','Bonferroni','FDR 25%'),col=1:3,lty=2)
```



## Linkage mapping

- Generally on experimental pops (F2, DH, RIL, BC)
- Based on single-marker analysis or interval mapping
- Recombination rates would increase **power**

![](https://www.jax.org/-/media/jaxweb/images/news-and-insights/articles/graphics/10-2004-lodvchrome.jpg)

## LD mapping (or association mapping)

- Use of historical LD between marker and QTL
- AM allowed studies on random panels
- Dense SNP panels would increase **resolution**

![](https://www.researchgate.net/profile/Charlotte_Acharya/publication/237198966/figure/fig11/AS:213410093506584@1427892339496/Genome-wide-association-study-GWAS-for-growing-degree-days-to-silking-GWAS-for-growing.png)

## Structure

1. Confounding associations with sub-populations
2. Major limitation of association mapping
3. Structure: *PCs*, *clusters*, *subpopulation* (eg. race)

```{r,echo=F,fig.width=8,fig.height=4,fig.align="center"}
pc = svd(gen,2,2)
ds = hclust(dist(gen),method = 'ward.D2')
k = cutree(ds,4)
#
par(mfrow = c(1,2),mar = c(1,1,3,0))
plot(ds,labels = F,axes=F); abline(h=50,lty=3,col=2)
plot(pc$u,pch=20,main='Principal Components',col=k,axes = F)
abline(h=0,v=-.07,lty=2,col=4)
```

## Structure

![](https://raw.githubusercontent.com/alenxav/miscellaneous/master/PC_Europe.png)

## Imputation

Less missing values = more obs. = more detection power

- Markov models: Based on flanking markers
- Random forest: Multiple decision trees capture LD
- kNN & Projections: Fill with similar haplotypes

![](https://raw.githubusercontent.com/alenxav/miscellaneous/master/Impute.jpg)

## GLM (generalized linear models)

- Full model ($L_1$): $$y = Xb + m_ja + e$$
- Null model ($L_0$): $$y = Xb + e$$

1. **Advantage**: Fast, not restricted to Gaussian traits
2. Popular methodology on human genetic studies
3. $Xb$ includes (1) environment, (2) structure and (3) covariates


## MLM (mixed linear models)

- Full model ($L_1$): $$y = Xb + Zu + m_ja + e$$
- Null model ($L_0$): $$y = Xb + Zu + e$$

1. The famous "**Q+K model**"
2. **Advantage**: Better control of false positives, no need for PCs
3. Polygenic effect ($u$) assumes $u\sim N(0,K\sigma^2_u)$
4. Faster if we don't reestimate $\lambda = \sigma^2_e/\sigma^2_u$ for each SNP


## cMLM (compressed MLM)

1. Uses the same base model as MLM
2. **Advantage**: Faster than MLM
3. Based on clustered individuals:
- $Z$ is indicates the subgroup
- $K$ is the relationship among subgroup
- Often needs PCs to complement $K$

## WGR (whole-genome regression)

1. Tests all markers at once
2. **Advantage**: No double-fitting, no PCs, no Bonferroni
- Model (BayesB, BayesC, SSVS): $$y = Xb + Ma + e$$ 
- Marker effects are from a mixture of distributions

$a_j \sim Binomial$ with $p(\pi) = 0$ and $p(1-\pi) = a_j$

## WGR (whole-genome regression)

```{r,echo=F,fig.width=8,fig.height=5,fig.align="center"}
mlm = emGWA(y,gen)
BB = emBB(y,gen,Pi=0.5)
par(mfrow=c(1,2))
plot(mlm$PVAL,main='MLM',ylab='-log(p)',xlab='SNP',type='h')
plot(-log(1-BB$d),main='BayesB',ylab='-log(p)',xlab='SNP',type='h')
```

## Rare variants

1. Screen a set ($s$) of low MAF markers on NGS data
2. **Advantage**: Detect signals from low power SNPs
3. Applied to uncommon diseases (not seen in plant breeding)
4. Two possible model
- Full model 1 ($L_1$): $y = Xb + M_sa + e$
- Full model 2 ($L_2$): $y = Xb + PC_1(M_s) + e$
- Null model ($L_0$): $y = Xb + e$

Test either $LR(L_1,L_0)$ or $LR(L_2,L_0)$

## Validation studies

- QTLs detected with 3 methods, across 3 mapping pops
- Validations made on 3 unrelated populations

![](https://raw.githubusercontent.com/alenxav/miscellaneous/master/QTL_validation.png)









